src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1120 @@
+    return variable ? variable.id : null;
+  }
+
+  function getListId(listName: string, target: Target, stage: Stage): string | null {
+    // Get the ID associated with the provided list name on either the provided
+    // target or the stage, or null if no matching list is found.
+
+    let list = target.getList(listName);
+    list = list || stage.getList(listName);
+    return list ? list.id : null;
+  }
+
+  interface SerializeInputsToFieldsOptions {
+    stage: Stage;
+    target: Target;
+
+    fieldEntries;
+  }
+
+  function serializeInputsToFields(
+    inputs: { [key: string]: BlockInput.Any },
+    options: SerializeInputsToFieldsOptions
+  ): {
+    fields: { [key: string]: any };
+  } {
+    // Serialize provided inputs into a "fields" mapping that can be stored
+    // on a serialized block.
+    //
+    // Where the Scratch 3.0 term "input" refers to a slot that accepts blocks,
+    // the term "field" is any user-interactive slot that cannot be obscured as
+    // such. Fields are also the interactive element within any shadow block,
+    // making their presence key to nearly all inputs.
+    //
+    // While inputs are fairly complex to serialize, fields are comparatively
+    // simple. A field always contains only its selected or entered value, so
+    // it's not concerned with a variety of structures like an input is.
+    // The format for a field mapping looks something like this:
+    //
+    //   {
+    //     VARIABLE: ["my variable", "theVariablesId"]
+    //   }
+    //
+    // Some fields take an ID; some don't. Here's another example:
+    //
+    //   {
+    //     DISTANCETOMENU: ["_mouse_"]
+    //   }
+    //
+    // The important thing to remember with fields during serialization is that
+    // they refer to slots that don't accept blocks: non-droppable menus,
+    // primarily, but the value in a (non-compressed) shadow input too.
 
-import List from "../../List";
-import Variable from "../../Variable";
+    const { fieldEntries, stage, target } = options;
 
-export default async function toSb3(project: Project): Promise<ArrayBuffer> {
-  const outZip = new JSZip();
+    const result = {
+      fields: {}
+    };
 
-  for (const target of [project.stage, ...project.sprites]) {
-    for (const costume of target.costumes) {
-      outZip.file(`${costume.md5}.${costume.ext}`, costume.asset);
+    if (!fieldEntries) {
+      return result;
     }
+
+    for (const key of Object.keys(fieldEntries)) {
+      const input = inputs[key];
+      // Fields are stored as a plain [value, id?] pair.
+      let id: string;
+      switch (input.type) {
+        case "variable":
+          id = getVariableId(input.value, target, stage);
+          break;
+        case "list":
+          id = getListId(input.value, target, stage);
+          break;
+        default:
+          id = null;
+          break;
+      }
+      result.fields[key] = [input.value, id];
+      continue;
```

Comment:
unnecessary `continue`?

Reply:
Apparently! It's been there since the first sb3 commit, so I have no idea how it ended up there. I'll get it removed :P